homeassistant/components/dyson/fan.py

```diff
@@ -10,62 +10,153 @@
+    vol.Required(ATTR_ANGLE_HIGH): cv.positive_int
+})
+
+DYSON_SET_FLOW_DIRECTION_FRONT_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Required(ATTR_FLOW_DIRECTION_FRONT): cv.boolean
+})
+
+DYSON_SET_TIMER_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Required(ATTR_TIMER): cv.positive_int
+})
+
+DYSON_SET_DYSON_SPEED_SCHEMA = vol.Schema({
+    vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+    vol.Required(ATTR_DYSON_SPEED): cv.positive_int
 })
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up the Dyson fan components."""
-    from libpurecoollink.dyson_pure_cool_link import DysonPureCoolLink
+    from libpurecool.dyson_pure_cool_link import DysonPureCoolLink
+    from libpurecool.dyson_pure_cool import DysonPureCool
+
+    if discovery_info is None:
+        return
 
     _LOGGER.debug("Creating new Dyson fans")
     if DYSON_FAN_DEVICES not in hass.data:
         hass.data[DYSON_FAN_DEVICES] = []
 
     # Get Dyson Devices from parent component
-    for device in [d for d in hass.data[DYSON_DEVICES] if
-                   isinstance(d, DysonPureCoolLink)]:
-        dyson_entity = DysonPureCoolLinkDevice(hass, device)
-        hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
+    has_purecool_devices = False
+    device_names = (device.name for device in hass.data[DYSON_FAN_DEVICES])
+    for device in hass.data[DYSON_DEVICES]:
+        if device.name not in device_names:
+            if isinstance(device, DysonPureCool):
+                has_purecool_devices = True
+                dyson_entity = DysonPureCoolDevice(device)
+                hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
+            elif isinstance(device, DysonPureCoolLink):
+                dyson_entity = DysonPureCoolLinkDevice(hass, device)
+                hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
 
     add_entities(hass.data[DYSON_FAN_DEVICES])
 
     def service_handle(service):
         """Handle the Dyson services."""
-        entity_id = service.data.get(CONF_ENTITY_ID)
-        night_mode = service.data.get(CONF_NIGHT_MODE)
-        fan_device = next([fan for fan in hass.data[DYSON_FAN_DEVICES] if
-                           fan.entity_id == entity_id].__iter__(), None)
+        entity_id = service.data[ATTR_ENTITY_ID]
+        night_mode = service.data[ATTR_NIGHT_MODE]
+        fan_device = next((fan for fan in hass.data[DYSON_FAN_DEVICES] if
+                           fan.entity_id == entity_id), None)
         if fan_device is None:
             _LOGGER.warning("Unable to find Dyson fan device %s",
                             str(entity_id))
             return
 
         if service.service == SERVICE_SET_NIGHT_MODE:
-            fan_device.night_mode(night_mode)
+            fan_device.set_night_mode(night_mode)
+
+        if has_purecool_devices:
+            auto_mode = service.data.get(ATTR_AUTO_MODE)
```

Comment:
Please go through all of the services in the service handler and update attribute access depending on service schema. Required options and options with default values don't need to use `dict.get`.

```suggestion
            auto_mode = service.data[ATTR_AUTO_MODE]
```

Reply:
Done.